### PR TITLE
Configurable storage engine for Netdata agents: step 3

### PR DESCRIPTION
**Pull request by aberaud**
Following the idea discussed here:
https://github.com/netdata/netdata/discussions/12216

This change introduces the new storage engine API for managing global and per-host storage engine contexts.

##### Summary

These commits allows storage engines to provide API functions to manage a global context and per-host contexts.
* In-memory engines don't use a context object. These engines provide a null function to signal that they don't use a context.
* dbengine uses a global context

##### Test Plan

These changes don't add any new memory modes/storage engines for now, and only refactor existing features, so they are expected to be covered by existing unit tests.

### CODE REVIEW

File: database/engine/rrdengine.c [hunk 1]
```diff
@@ -1339,13 +1339,12 @@ void rrdeng_worker(void* arg)
...
 {
-    int ret;
-    struct rrdengine_instance *ctx;
+    STORAGE_ENGINE_INSTANCE* ctx;
 
     sanity_check();
-    ret = rrdeng_init(NULL, &ctx, "/tmp", RRDENG_MIN_PAGE_CACHE_SIZE_MB, RRDENG_MIN_DISK_SPACE_MB);
-    if (ret) {
-        exit(ret);
+    ctx = rrdeng_init(storage_engine_get(RRD_MEMORY_MODE_DBENGINE), NULL);
```
Comment (thiagoftsm) [inline database/engine/rrdengine.c:1345]: Hello @aberaud ,

The `rrdeng_init` function is calling [callocz](https://github.com/netdata/netdata/blob/f62bc0efddb2a5ddaa11655bb59771478b11e44b/libnetdata/libnetdata.c#L175-L179) that will close netdata calling `fatal` when memory cannot be allocated, so next test calling `exit` is not necessary.

Comment (aberaud) [reply database/engine/rrdengine.c]: Done

**Review by thiagoftsm — APPROVED** [review]
After last commits I tested this PR using `memory mode` with values `dbengine`, `ram`, and `map`. I did not observe any problem to access local dashboard, cloud, and streaming. I also did not have any error.

**Comment by underhood** [comment]
this needs rebase after recent rrdcontext changes merge

**Comment by stelfrag** [comment]
N/A anymore due to code changes

**Pull request closed without merging**